include/gatekeeper_lpm.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_LPM_H_
+#define _GATEKEEPER_LPM_H_
+
+#include <rte_lpm.h>
+#include <rte_lpm6.h>
+#include <rte_atomic.h>
+
+struct ipv4_lpm_route {
+	uint32_t ip;
+	uint8_t  depth;
+	uint8_t  policy_id;
+};
+
+struct ipv6_lpm_route {
+	uint8_t ip[16];
+	uint8_t depth;
+	uint8_t policy_id;
+};
+
+struct lpm_rt {
+	char            tag[128];
+	/*
+	 * For ease of update, each structure should maintain
+	 * two IPv4 and two IPv6 LPM tables:
+	 * one is for lookup, and one is for update.
+	 */
+	struct rte_lpm  *ipv4_rt[2];
+	struct rte_lpm6 *ipv6_rt[2];
```

Comment:
Yes, I have read the code. The LPM uses 1-bit ``valid flag`` field to determine whether the entry/table is valid or not. The functions ``rte_lpm_add()`` and ``rte_lpm_delete()`` will set this field when do the update, and the lookup function will first check this field to determine the next step. However, there is no protection on this field among these readers and writers. Maybe we need to make a trade off here, if update is infrequent, and we allow some potential minor inconsistency, we can simplify the code greatly. If the update is very frequent, we may need to consider this effect. I think the reader is the thread who performs the lookup function on the LPM table, while the writer is the dynamic configuration block.